certora/specs/ERC20Invariants.spec

```diff
@@ -0,0 +1,138 @@
+hook Sload uint256 balance (slot 0x52c63247e1f47db19d5ce0460030c497f067ca4cebf71ba98eeadabe20bace00)[KEY address account] {
+    require ghostBalances[account] == balance;
+}
+
+// Slot is ERC20Storage._balances slot
+hook Sstore (slot 0x52c63247e1f47db19d5ce0460030c497f067ca4cebf71ba98eeadabe20bace00)[KEY address account] uint256 newValue (uint256 oldValue) {
+    // Update partial sum of balances, for x > to_mathint(account)
+    // Track balance changes in balances.
+    havoc sumOfBalances assuming
+        forall mathint x. sumOfBalances@new[x] ==
+            sumOfBalances@old[x] + (to_mathint(account) < x ? newValue - oldValue : 0);
+    // Update partial sums of votes delegated to the parameterized address, for x > to_mathint(account)
+    // Track balance changes when the delegatee is the parameterized address.
+    if (ghostDelegatee[account] == A) {
+        havoc sumsOfVotes assuming
+            forall mathint x. sumsOfVotes@new[x] ==
+                sumsOfVotes@old[x] + (to_mathint(account) < x ? newValue - oldValue : 0);
+    }
+    // Update ghost copy of ERC20Storage._balances.
+    ghostBalances[account] = newValue;
+}
+
+invariant sumOfBalancesStartsAtZero()
+    sumOfBalances[0] == 0;
+
+invariant sumOfBalancesGrowsCorrectly()
+    forall address addr. sumOfBalances[to_mathint(addr) + 1] ==
+        sumOfBalances[to_mathint(addr)] + ghostBalances[addr];
+
+invariant sumOfBalancesMonotone()
+    forall mathint i. forall mathint j. i <= j => sumOfBalances[i] <= sumOfBalances[j]
+    {
+        preserved {
+            requireInvariant sumOfBalancesStartsAtZero();
+            requireInvariant sumOfBalancesGrowsCorrectly();
+        }
+    }
+
+// Check that the sum of balances equals the total supply.
+invariant totalSupplyIsSumOfBalances()
+    sumOfBalances[2^160] == to_mathint(totalSupply())
+    {
+        preserved {
+            requireInvariant sumOfBalancesStartsAtZero();
+            requireInvariant sumOfBalancesGrowsCorrectly();
+            requireInvariant sumOfBalancesMonotone();
+        }
+    }
+
+invariant balancesLTEqTotalSupply()
+    forall address a. ghostBalances[a] <= sumOfBalances[2^160]
+    {
+        preserved {
+            requireInvariant sumOfBalancesStartsAtZero();
+            requireInvariant sumOfBalancesGrowsCorrectly();
+            requireInvariant sumOfBalancesMonotone();
+            requireInvariant totalSupplyIsSumOfBalances();
+        }
+    }
+
+rule twoBalancesCannotExceedTotalSupply(address accountA, address accountB) {
+    requireInvariant sumOfBalancesStartsAtZero();
+    requireInvariant sumOfBalancesGrowsCorrectly();
+    requireInvariant sumOfBalancesMonotone();
+    requireInvariant totalSupplyIsSumOfBalances();
+    uint256 balanceA = balanceOf(accountA);
+    uint256 balanceB = balanceOf(accountB);
+
+    assert accountA != accountB =>
+        balanceA + balanceB <= to_mathint(totalSupply());
+    satisfy(accountA != accountB && balanceA > 0 && balanceB > 0);
```

Comment:
why do we need this

Reply:
This is to ensure that the rule is no trivially true by requiring to produce a witness where balances are greater than 0.

Reply:
but accountB and accountA are chosen arbitrarily? so it's like saying at least two people can have a non nul balance?

Reply:
`satisfy` can be understood as, from the current state are we **able** to show the following, in contrast with `assert`, which says ensure that this is **always** true. So in this case we are no restricting the the proof to be true only if `accountA` and `accountB` have non zero balances but merely, now that we have proven that our lemma is true, can we derive a case that is true such that `accountA` and `accountB` have non zero balances. Is that helpful?

Reply:
Note also, that in https://github.com/morpho-org/morpho-token/pull/93/commits/cf1b556b13e9517ecae13fe4c23a0e8a530b6610 this changed to invariant.